File: mahaam-api-ts/src/infra/rule.ts
/**
 * Validation utility functions
 */

import { InputError } from './errors';

export function required(value: string | null | undefined, name: string): void {
  if (!value || value.trim() === '') {
    throw new InputError(`${name} is required`);
  }
}

export function oneAtLeastRequired(values: (any | null | undefined)[], message: string): void {
  const hasValidValue = values.some((item) => item !== null && item !== undefined && (typeof item !== 'string' || item.trim() !== ''));

  if (!hasValidValue) {
    throw new InputError(message);
  }
}

export function requiredGuid(value: string | null | undefined, name: string): void {
  if (!value || value.trim() === '' || value === '00000000-0000-0000-0000-000000000000') {
    throw new InputError(`${name} is required`);
  }
}

export function requiredBoolean(value: boolean | null | undefined, name: string): void {
  if (value === null || value === undefined) {
    throw new InputError(`${name} is required`);
  }
}

export function requiredNumber(value: number | null | undefined, name: string): void {
  if (value === null || value === undefined) {
    throw new InputError(`${name} is required`);
  }
}

export function requiredObject(value: any | null | undefined, name: string): void {
  if (value === null || value === undefined) {
    throw new InputError(`${name} is required`);
  }
}

export function isIn(item: string, list: string[]): void {
  if (!list.includes(item)) {
    throw new InputError(`${item} is not in [${list.join(',')}]`);
  }
}

export function failIf(condition: boolean, message: string): void {
  if (condition) {
    throw new InputError(message);
  }
}

export function validateEmail(email: string): void {
  required(email, 'email');

  // Email validation regex pattern
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

  if (!emailRegex.test(email)) {
    throw new InputError('Invalid email');
  }
}

File: mahaam-api-java/src/main/java/mahaam/infra/Rule.java
package mahaam.infra;

import java.util.List;
import java.util.UUID;
import java.util.regex.Pattern;

import mahaam.infra.Exceptions.InputException;

public class Rule {

	private static final Pattern EMAIL_PATTERN = Pattern.compile(
			"^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$");

	public static void required(String value, String name) {
		if (value == null || value.trim().isEmpty())
			throw new InputException(name + " is required");
	}

	public static void oneAtLeastRequired(List<?> values, String message) {
		if (values.stream().allMatch(item -> item == null || item.toString().trim().isEmpty()))
			throw new InputException(message);
	}

	public static void required(UUID value, String name) {
		if (value == null)
			throw new InputException(name + " is required");
	}

	public static void required(Boolean value, String name) {
		if (value == null)
			throw new InputException(name + " is required");
	}

	public static void required(Integer value, String name) {
		if (value == null)
			throw new InputException(name + " is required");
	}

	public static void required(Object value, String name) {
		if (value == null)
			throw new InputException(name + " is required");
	}

	public static void in(String item, List<String> list) {
		if (!list.contains(item))
			throw new InputException(item + " is not in [" + String.join(",", list) + "]");
	}

	public static void failIf(boolean condition, String message) {
		if (condition)
			throw new InputException(message);
	}

	public static void validateEmail(String email) {
		required(email, "email");
		if (!EMAIL_PATTERN.matcher(email).matches())
			throw new InputException("Invalid email");
	}
}
